src/qibolab/instruments/qm/program/instructions.py

```diff
@@ -6,23 +6,30 @@
 
 from qibolab.components import Config
 from qibolab.execution_parameters import AcquisitionType, ExecutionParameters
-from qibolab.pulses import Delay, Pulse
-from qibolab.sequence import PulseSequence
+from qibolab.pulses import Align, Delay, Pulse, VirtualZ
 from qibolab.sweeper import ParallelSweepers
 
 from ..config import operation
-from .acquisition import Acquisition, Acquisitions
+from .acquisition import Acquisition
 from .arguments import ExecutionArguments, Parameters
-from .sweepers import INT_TYPE, NORMALIZERS, SWEEPER_METHODS
+from .sweepers import INT_TYPE, NORMALIZERS, SWEEPER_METHODS, normalize_phase
 
 
 def _delay(pulse: Delay, element: str, parameters: Parameters):
     # TODO: How to play delays on multiple elements?
     if parameters.duration is None:
-        duration = int(pulse.duration) // 4 + 1
+        duration = int(pulse.duration) // 4
     else:
         duration = parameters.duration
-    qua.wait(duration, element)
+    qua.wait(duration + 1, element)
+
+
+def _play_multiple_waveforms(element: str, parameters: Parameters):
+    """Sweeping pulse duration using distinctly uploaded waveforms."""
+    with qua.switch_(parameters.duration, unsafe=True):
+        for value, sweep_op in parameters.pulses:
+            with qua.case_(value // 4):
```

Comment:
I guess this is to avoid baking. What happens for the values covered multiple times?

Reply:
And isn't there a `// 4` already in the normalizing function?

Reply:
For now this is not related to baking, but it will most likely need to be modified when baking is implemented.

Basically, the reason we need the `// 4` in `case_` is because it also exists in the normalizing function. In particular, the loop variable (`parameters.duration`) will take values from an array that was normalized with `// 4`, while the uploaded waveforms are enumerated with the original array, so we have to divide the enumerator when we do the matching, in order to properly match with the loop variable. We could drop the normalization in both places, however then we would need to make sure to normalize when delays are involved, because [`qua.wait`](https://docs.quantum-machines.co/1.2.0/docs/API_references/qua/dsl_main/#qm.qua._dsl.wait) expects clock cycles (=4ns).

Anyway, this will probably change when baking (=supporting pulses with 1ns resolution, the best we can currently do) is implemented, so I would postpone until then, as I am planning to do it right after this PR anyway, together with the other issues from #969. Otherwise, I could push baking directly here, as I don't expect it to be very complicated.

Reply:
Ok, so I get that if you're receiving as values `[4, 5, 6, 7]` you'll sweep four elements, but the value will always be the same one.

For as long as it is documented, for me it's perfectly acceptable.

Reply:
Indeed, these values are not supported yet since baking is not implemented, so only waveforms with length that is multiple of 4 are supported (and >16). I am not sure what will be the exact behavior as I haven't tried, but I would guess that an error will appear if we try to upload a waveform of these lengths. In any case, I will implement this before we release 0.2 so it should be fine. For now I added a reference to this discussion in #969.

One additional point to check is what happens to the interpolated sweep in this case. I guess for these lengths it will fail because they are very short, however for something like [20, 21, 22, 23] it may not fail and actually return four overlapping points. If that happens we may want to raise an error ourselves and prompt to use non-interpolated sweep (which will support baking).